awx/main/models/events.py

```diff
@@ -575,12 +575,26 @@ def _update_host_summary_from_stats(self, hostnames):
 
             Host.objects.bulk_update(list(updated_hosts), ['last_job_id', 'last_job_host_summary_id'], batch_size=100)
 
-            # bulk-create
-            current_time = now()
-            HostMetric.objects.bulk_create(
-                [HostMetric(hostname=hostname, last_automation=current_time) for hostname in updated_hosts_list], ignore_conflicts=True, batch_size=100
+            # Create/update Host Metrics
+            self._update_host_metrics(updated_hosts_list)
+
+    @staticmethod
+    def _update_host_metrics(updated_hosts_list):
+        from awx.main.models import HostMetric  # circular import
+
+        # bulk-create
+        current_time = now()
+        HostMetric.objects.bulk_create(
+            [HostMetric(hostname=hostname, last_automation=current_time) for hostname in updated_hosts_list], ignore_conflicts=True, batch_size=1000
+        )
+        # bulk-update
+        batch_start, batch_size = 0, 1000
+        while batch_start <= len(updated_hosts_list):
+            batched_host_list = updated_hosts_list[batch_start : (batch_start + batch_size)]
+            HostMetric.objects.filter(hostname__in=batched_host_list).update(
+                last_automation=current_time, automated_counter=models.F('automated_counter') + 1, deleted=False
             )
-            HostMetric.objects.filter(hostname__in=updated_hosts_list).update(last_automation=current_time)
+            batch_start += batch_size
```

Comment:
@kdelee @AlanCoding do you have any concern about the performance implication of this change?

Reply:
I'd ballpark guesstimate that it would take 2x as long as it did before. Instead of bulk creating these, now this will bulk-create and then bulk-update. While that's not ideal, I believe that shortcoming was obvious from the get-go and this is a price paid to keep the operations in the domain of the ORM, which I agree with.

I would like to use the Django `batch_size` parameter as opposed to a hard-coded batch size, but I realize that's not compatible with `.update`, because the number of rows affect is unknown by Django (although known by us).

This does lead us into the obvious potential improvement, which is that we could have this python code apply additional fields like `last_automation` to the `.bulk_create` and exclude those hosts from the `.update`. Even if that includes some unlikely race conditions I think it would be my preference, but this is getting too deep into things. What we have here should probably suffice.

Reply:
I tried to use upsert first, but it's not implemented in Django's ORM. And using psycopg2 with upsert is not compatible with sqlite3 in tests. 
Anyway yes, if the `bulk_create` returns created hostnames without ignored, I can exclude them from `update`

Reply:
@AlanCoding the update was there even before, so this shouldn't be adding more processing time.

Martin is adding batching for the update, because before it wasn't batched. Afaik it should be batched because from my experience doing a query with "hostname IN <100k hostnames>" will just explode in DB. (mostly because PG will not do index scan) So this approach should be faster for jobs with a lot of hosts.

-------

Martin already has the upsert query (doing atomic insert and update as one query) which will be much more performant, but we're not able to test it with SQLite. (I was looking for a while how to make some tests run with Postgres in AWX, but it doesn't seems that trivial to do. Although e.g. Automation Analytics runs tests on Postgres)

Reply:
@AlanCoding 
I tested the bulk_create:
In Django 3.2 (used in awx), the bulk_create returns all objects including ignored. And even ignored have updated values. So it's not possible to recognize what was and what wasn't updated without additional query => excluding from update is not worth. 
Regarding [django 4.2 doc](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#bulk-create) the `bulk_create` offers `update_fields` and `update_conflicts` (not clear about effectivity). So in order to keep usage of ORM and be more effective Django update is required

Reply:
Thanks, those are great replies. I believe this is a carefully engineered set of queries, and I'm not particularly worried about performance.

One thing that is truly novel here is the 1000 manual batch size. The Host `name` field has a max of 512 characters. Are Django and postgres okay if you run a job with >1000 hosts where they all have a name 512 characters? That's the main risk I can see being introduced here, and could be handled with a fairly straightforward test case.

Reply:
I changed batch_size from 100 -> 1000 in `bulk_create`, but decreased from 'unlimited' -> 1000 in `update`. 
So I can decrease back the `bulk_create`, it's the main concern of @AlanCoding if I understand correctly. 
This optimization is not directly subject of this PR. Moreover this testing can't be done in this repo's tests (because performance testing against sqlite3 doesn't make much sense)